File: api/library/conflicts.py
# * =========================
# * NEGATIVE CONFLICT PATTERNS (DERIVED FROM CORE)
# * =========================
# All negative/destructive branch and stem interactions
# Patterns are DERIVED from core.py STEMS and BRANCHES.

from .core import STEMS, BRANCHES
from .derived import ELEMENT_CYCLES
from .scoring import (
    BASE_SCORES,
    generate_single_scoring,
    generate_asymmetric_scoring,
)

# * -------------------------
# * SCORING CONFIGURATIONS
# * -------------------------

_CLASHES_OPPOSITE_SCORING = generate_asymmetric_scoring(
    BASE_SCORES["CLASHES_OPPOSITE"]["damage"],
    "two_branch",
    ratio=0.618
)
_CLASHES_SAME_SCORING = generate_single_scoring(
    BASE_SCORES["CLASHES_SAME"]["damage"],
    "two_branch"
)
_STEM_CONFLICTS_SCORING = generate_asymmetric_scoring(
    BASE_SCORES["STEM_CONFLICTS"]["damage"],
    "two_branch",
    ratio=0.618
)
_PUNISHMENTS_3NODE_SCORING = generate_single_scoring(
    BASE_SCORES["PUNISHMENTS_3NODE"]["damage"],
    "two_branch"
)
_PUNISHMENTS_2NODE_SCORING = generate_asymmetric_scoring(
    BASE_SCORES["PUNISHMENTS_2NODE"]["damage"],
    "two_branch",
    ratio=0.618
)
_PUNISHMENTS_SELF_SCORING = generate_single_scoring(
    BASE_SCORES["PUNISHMENTS_2NODE"]["damage"],
    "two_branch"
)
_HARMS_SCORING = generate_asymmetric_scoring(
    BASE_SCORES["HARMS"]["damage"],
    "two_branch",
    ratio=0.618
)
_DESTRUCTION_OPPOSITE_SCORING = generate_asymmetric_scoring(
    BASE_SCORES["DESTRUCTION_OPPOSITE"]["damage"],
    "two_branch",
    ratio=0.618
)
_DESTRUCTION_SAME_SCORING = generate_single_scoring(
    BASE_SCORES["DESTRUCTION_SAME"]["damage"],
    "two_branch"
)

# * -------------------------
# * PUNISHMENTS (刑)
# * -------------------------

_punishment_groups = {}
_self_punishments = []

for branch_id, branch in BRANCHES.items():
    punishment_group = branch.get("punishment_group")
    if punishment_group:
        group_key = "-".join(punishment_group[:3])
        if group_key not in _punishment_groups:
            _punishment_groups[group_key] = {
                "branches": list(punishment_group[:3]),
                "type": punishment_group[3]
            }

    punishment_pair = branch.get("punishment_pair")
    if punishment_pair:
        partner, pun_type, role = punishment_pair
        pair_key = f"{branch_id}-{partner}" if role == "controller" else f"{partner}-{branch_id}"
        if pair_key not in _punishment_groups:
            _punishment_groups[pair_key] = {
                "branches": [branch_id, partner] if role == "controller" else [partner, branch_id],
                "type": pun_type,
                "controller": branch_id if role == "controller" else partner,
                "controlled": partner if role == "controller" else branch_id,
            }

    if branch.get("self_punishment"):
        _self_punishments.append(branch_id)

PUNISHMENTS = {}

for group_key, group in _punishment_groups.items():
    branches = group["branches"]
    pun_type = group["type"]

    if len(branches) == 3:
        if pun_type == "shi":
            PUNISHMENTS[group_key] = {
                "branches": branches,
                "type": "3-node",
                "category": "shi_xing",
                "chinese_name": "勢刑",
                "english_name": "Bullying Punishment",
                "severity": "severe",
                "element_conflict": True,
                "scoring": _PUNISHMENTS_3NODE_SCORING,
                "description": "Three powerful branches in mutual conflict (恃勢之刑)"
            }
        elif pun_type == "wu_li":
            PUNISHMENTS[group_key] = {
                "branches": branches,
                "type": "3-node",
                "category": "wu_li_xing",
                "chinese_name": "無禮刑",
                "english_name": "Rudeness Punishment",
                "severity": "moderate",
                "element_conflict": False,
                "scoring": _PUNISHMENTS_3NODE_SCORING,
                "description": "Earth branches showing disrespect to each other (無禮之刑)"
            }
    elif len(branches) == 2 and pun_type == "en":
        controller = group.get("controller")
        controlled = group.get("controlled")
        PUNISHMENTS[group_key] = {
            "branches": branches,
            "type": "2-node",
            "category": "en_xing",
            "chinese_name": "恩刑",
            "english_name": "Ungrateful Punishment",
            "severity": "light",
            "controller": controller,
            "controlled": controlled,
            "element_relationship": "productive",
            "scoring": _PUNISHMENTS_2NODE_SCORING,
            "description": "Beneficiary punishes benefactor - betrayal (持恩之刑)"
        }

# Self-punishment data now embedded in BRANCHES as self_punishment_nature
for branch_id in _self_punishments:
    branch = BRANCHES[branch_id]
    nature = branch.get("self_punishment_nature", "Self-conflict")
    key = f"{branch_id}-{branch_id}"
    PUNISHMENTS[key] = {
        "branches": [branch_id, branch_id],
        "type": "self",
        "category": "zi_xing",
        "chinese_name": "自刑",
        "english_name": f"Self-Punishment ({branch['animal']})",
        "severity": "self",
        "element": branch["element"],
        "nature": nature,
        "scoring": _PUNISHMENTS_SELF_SCORING,
        "description": f"{branch['animal']}'s nature punishes itself ({branch_id}自刑)"
    }

PUNISHMENT_SEVERITY = {
    "severe": 1.0,
    "moderate": 0.85,
    "light": 0.70,
    "self": 0.60
}

# * -------------------------
# * HARMS (害)
# * -------------------------

_harms_seen = set()
HARMS = {}
for branch_id, branch in BRANCHES.items():
    partner = branch.get("harms")
    role = branch.get("harm_role")
    if partner and role:
        pair = tuple(sorted([branch_id, partner]))
        if pair not in _harms_seen:
            _harms_seen.add(pair)
            if role == "controller":
                controller, controlled = branch_id, partner
            else:
                controller, controlled = partner, branch_id
            key = f"{pair[0]}-{pair[1]}"
            HARMS[key] = {
                "branches": list(pair),
                "controller": controller,
                "controlled": controlled,
                "scoring": _HARMS_SCORING
            }

# * -------------------------
# * CLASHES (沖)
# * -------------------------

_clashes_seen = set()
CLASHES = {}
for branch_id, branch in BRANCHES.items():
    partner = branch.get("clashes")
    if partner:
        pair = tuple(sorted([branch_id, partner]))
        if pair not in _clashes_seen:
            _clashes_seen.add(pair)
            b1_elem = branch["element"]
            b2_elem = BRANCHES[partner]["element"]

            if b1_elem == b2_elem:
                clash_type = "same"
                scoring = _CLASHES_SAME_SCORING
                key = f"{pair[0]}-{pair[1]}"
                CLASHES[key] = {
                    "branches": list(pair),
                    "type": clash_type,
                    "scoring": scoring
                }
            else:
                clash_type = "opposite"
                scoring = _CLASHES_OPPOSITE_SCORING
                if ELEMENT_CYCLES["controlling"].get(b1_elem) == b2_elem:
                    controller, controlled = branch_id, partner
                else:
                    controller, controlled = partner, branch_id
                key = f"{pair[0]}-{pair[1]}"
                CLASHES[key] = {
                    "branches": list(pair),
                    "type": clash_type,
                    "controller": controller,
                    "controlled": controlled,
                    "scoring": scoring
                }

# * -------------------------
# * DESTRUCTION (破)
# * -------------------------

_destruction_seen = set()
DESTRUCTION = {}
for branch_id, branch in BRANCHES.items():
    partner = branch.get("destroys")
    if partner:
        pair = tuple(sorted([branch_id, partner]))
        if pair not in _destruction_seen:
            _destruction_seen.add(pair)
            b1_elem = branch["element"]
            b2_elem = BRANCHES[partner]["element"]

            if b1_elem == b2_elem:
                dest_type = "same"
                scoring = _DESTRUCTION_SAME_SCORING
                key = f"{pair[0]}-{pair[1]}"
                DESTRUCTION[key] = {
                    "branches": list(pair),
                    "type": dest_type,
                    "scoring": scoring
                }
            else:
                dest_type = "opposite"
                scoring = _DESTRUCTION_OPPOSITE_SCORING
                if ELEMENT_CYCLES["generating"].get(b1_elem) == b2_elem:
                    controller, controlled = branch_id, partner
                else:
                    controller, controlled = partner, branch_id
                key = f"{pair[0]}-{pair[1]}"
                DESTRUCTION[key] = {
                    "branches": list(pair),
                    "type": dest_type,
                    "controller": controller,
                    "controlled": controlled,
                    "scoring": scoring
                }

# * -------------------------
# * STEM CONFLICTS (天干相剋)
# * -------------------------

_stem_conflicts_seen = set()
STEM_CONFLICTS = {}
for stem_id, stem in STEMS.items():
    controlled = stem.get("controls")
    if controlled:
        pair = tuple(sorted([stem_id, controlled]))
        if pair not in _stem_conflicts_seen:
            _stem_conflicts_seen.add(pair)
            key = f"{pair[0]}-{pair[1]}"
            STEM_CONFLICTS[key] = {
                "controller": stem_id,
                "controlled": controlled,
                "scoring": _STEM_CONFLICTS_SCORING
            }

# * -------------------------
# * WUXING ENERGY FLOW (五行流動)
# * -------------------------

WUXING_ENERGY_FLOW = {
    "generation": ELEMENT_CYCLES["generating"],
    "control": ELEMENT_CYCLES["controlling"],
    "scoring": {
        "generating": {
            "generator_loss": 7,
            "receiver_gain": 10,
            "count_increment": 0.25
        },
        "controlling": {
            "controller_loss": 5,
            "controlled_loss": 10,
            "count_decrement": 0.25
        }
    }
}

File: src/lib/bazi/conflicts.ts
import 'server-only';

// =============================================================================
// NEGATIVE CONFLICT PATTERNS (DERIVED FROM CORE)
// =============================================================================
// All negative/destructive branch and stem interactions
// Patterns are DERIVED from core.ts STEMS and BRANCHES.

import { STEMS, BRANCHES, type StemName, type BranchName, type BranchData, type Element } from './core';
import { ELEMENT_CYCLES } from './derived';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface PunishmentEntry {
  readonly branches: readonly string[];
  readonly type: "3-node" | "2-node" | "self";
  readonly category: string;
  readonly chinese_name: string;
  readonly english_name: string;
  readonly severity: "severe" | "moderate" | "light" | "self";
  readonly element_conflict?: boolean;
  readonly controller?: string;
  readonly controlled?: string;
  readonly element_relationship?: string;
  readonly element?: Element;
  readonly nature?: string;
  readonly description: string;
}

export interface HarmEntry {
  readonly branches: readonly string[];
  readonly controller: string;
  readonly controlled: string;
}

export interface ClashEntry {
  readonly branches: readonly string[];
  readonly type: "same" | "opposite";
  readonly controller?: string;
  readonly controlled?: string;
}

export interface DestructionEntry {
  readonly branches: readonly string[];
  readonly type: "same" | "opposite";
  readonly controller?: string;
  readonly controlled?: string;
}

export interface StemConflictEntry {
  readonly controller: string;
  readonly controlled: string;
}

// ---------------------------------------------------------------------------
// PUNISHMENTS (刑)
// ---------------------------------------------------------------------------

interface PunishmentGroupRaw {
  branches: string[];
  type: string;
  controller?: string;
  controlled?: string;
}

const _punishmentGroups: Record<string, PunishmentGroupRaw> = {};
const _selfPunishments: BranchName[] = [];

for (const branchId of Object.keys(BRANCHES) as BranchName[]) {
  const branch = BRANCHES[branchId] as BranchData;

  const punishmentGroup = branch.punishment_group;
  if (punishmentGroup) {
    const groupKey = `${punishmentGroup[0]}-${punishmentGroup[1]}-${punishmentGroup[2]}`;
    if (!(groupKey in _punishmentGroups)) {
      _punishmentGroups[groupKey] = {
        branches: [punishmentGroup[0], punishmentGroup[1], punishmentGroup[2]],
        type: punishmentGroup[3],
      };
    }
  }

  const punishmentPair = branch.punishment_pair;
  if (punishmentPair) {
    const [partner, punType, role] = punishmentPair;
    const pairKey = role === "controller"
      ? `${branchId}-${partner}`
      : `${partner}-${branchId}`;
    if (!(pairKey in _punishmentGroups)) {
      _punishmentGroups[pairKey] = {
        branches: role === "controller"
          ? [branchId, partner]
          : [partner, branchId],
        type: punType,
        controller: role === "controller" ? branchId : partner,
        controlled: role === "controller" ? partner : branchId,
      };
    }
  }

  if (branch.self_punishment) {
    _selfPunishments.push(branchId);
  }
}

const _punishments: Record<string, PunishmentEntry> = {};

for (const [groupKey, group] of Object.entries(_punishmentGroups)) {
  const branches = group.branches;
  const punType = group.type;

  if (branches.length === 3) {
    if (punType === "shi") {
      _punishments[groupKey] = {
        branches,
        type: "3-node",
        category: "shi_xing",
        chinese_name: "勢刑",
        english_name: "Bullying Punishment",
        severity: "severe",
        element_conflict: true,
        description: "Three powerful branches in mutual conflict (恃勢之刑)",
      };
    } else if (punType === "wu_li") {
      _punishments[groupKey] = {
        branches,
        type: "3-node",
        category: "wu_li_xing",
        chinese_name: "無禮刑",
        english_name: "Rudeness Punishment",
        severity: "moderate",
        element_conflict: false,
        description: "Earth branches showing disrespect to each other (無禮之刑)",
      };
    }
  } else if (branches.length === 2 && punType === "en") {
    const controller = group.controller!;
    const controlled = group.controlled!;
    _punishments[groupKey] = {
      branches,
      type: "2-node",
      category: "en_xing",
      chinese_name: "恩刑",
      english_name: "Ungrateful Punishment",
      severity: "light",
      controller,
      controlled,
      element_relationship: "productive",
      description: "Beneficiary punishes benefactor - betrayal (持恩之刑)",
    };
  }
}

// Self-punishment data now embedded in BRANCHES as self_punishment_nature
for (const branchId of _selfPunishments) {
  const branch = BRANCHES[branchId] as BranchData;
  const nature = branch.self_punishment_nature ?? "Self-conflict";
  const key = `${branchId}-${branchId}`;
  _punishments[key] = {
    branches: [branchId, branchId],
    type: "self",
    category: "zi_xing",
    chinese_name: "自刑",
    english_name: `Self-Punishment (${branch.animal})`,
    severity: "self",
    element: branch.element,
    nature,
    description: `${branch.animal}'s nature punishes itself (${branchId}自刑)`,
  };
}

export const PUNISHMENTS: Readonly<Record<string, PunishmentEntry>> = _punishments;

export const PUNISHMENT_SEVERITY: Readonly<Record<string, number>> = {
  severe: 1.0,
  moderate: 0.85,
  light: 0.70,
  self: 0.60,
};

// ---------------------------------------------------------------------------
// HARMS (害)
// ---------------------------------------------------------------------------

const _harmsSeen = new Set<string>();
const _harms: Record<string, HarmEntry> = {};
for (const branchId of Object.keys(BRANCHES) as BranchName[]) {
  const branch = BRANCHES[branchId];
  const partner = branch.harms;
  const role = branch.harm_role;
  if (partner && role) {
    const pair = [branchId, partner].sort() as [string, string];
    const pairKey = pair.join(",");
    if (!_harmsSeen.has(pairKey)) {
      _harmsSeen.add(pairKey);
      let controller: string;
      let controlled: string;
      if (role === "controller") {
        controller = branchId;
        controlled = partner;
      } else {
        controller = partner;
        controlled = branchId;
      }
      const key = `${pair[0]}-${pair[1]}`;
      _harms[key] = {
        branches: pair,
        controller,
        controlled,
      };
    }
  }
}
export const HARMS: Readonly<Record<string, HarmEntry>> = _harms;

// ---------------------------------------------------------------------------
// CLASHES (沖)
// ---------------------------------------------------------------------------

const _clashesSeen = new Set<string>();
const _clashes: Record<string, ClashEntry> = {};
for (const branchId of Object.keys(BRANCHES) as BranchName[]) {
  const branch = BRANCHES[branchId];
  const partner = branch.clashes;
  if (partner) {
    const pair = [branchId, partner].sort() as [string, string];
    const pairKey = pair.join(",");
    if (!_clashesSeen.has(pairKey)) {
      _clashesSeen.add(pairKey);
      const b1Elem = branch.element;
      const b2Elem = BRANCHES[partner].element;
      const key = `${pair[0]}-${pair[1]}`;

      if (b1Elem === b2Elem) {
        _clashes[key] = {
          branches: pair,
          type: "same",
        };
      } else {
        let controller: string;
        let controlled: string;
        if (ELEMENT_CYCLES.controlling[b1Elem] === b2Elem) {
          controller = branchId;
          controlled = partner;
        } else {
          controller = partner;
          controlled = branchId;
        }
        _clashes[key] = {
          branches: pair,
          type: "opposite",
          controller,
          controlled,
        };
      }
    }
  }
}
export const CLASHES: Readonly<Record<string, ClashEntry>> = _clashes;

// ---------------------------------------------------------------------------
// DESTRUCTION (破)
// ---------------------------------------------------------------------------

const _destructionSeen = new Set<string>();
const _destruction: Record<string, DestructionEntry> = {};
for (const branchId of Object.keys(BRANCHES) as BranchName[]) {
  const branch = BRANCHES[branchId];
  const partner = branch.destroys;
  if (partner) {
    const pair = [branchId, partner].sort() as [string, string];
    const pairKey = pair.join(",");
    if (!_destructionSeen.has(pairKey)) {
      _destructionSeen.add(pairKey);
      const b1Elem = branch.element;
      const b2Elem = BRANCHES[partner].element;
      const key = `${pair[0]}-${pair[1]}`;

      if (b1Elem === b2Elem) {
        _destruction[key] = {
          branches: pair,
          type: "same",
        };
      } else {
        let controller: string;
        let controlled: string;
        if (ELEMENT_CYCLES.generating[b1Elem] === b2Elem) {
          controller = branchId;
          controlled = partner;
        } else {
          controller = partner;
          controlled = branchId;
        }
        _destruction[key] = {
          branches: pair,
          type: "opposite",
          controller,
          controlled,
        };
      }
    }
  }
}
export const DESTRUCTION: Readonly<Record<string, DestructionEntry>> = _destruction;

// ---------------------------------------------------------------------------
// STEM CONFLICTS (天干相剋)
// ---------------------------------------------------------------------------

const _stemConflictsSeen = new Set<string>();
const _stemConflicts: Record<string, StemConflictEntry> = {};
for (const stemId of Object.keys(STEMS) as StemName[]) {
  const stem = STEMS[stemId];
  const controlled = stem.controls;
  if (controlled) {
    const pair = [stemId, controlled].sort() as [string, string];
    const pairKey = pair.join(",");
    if (!_stemConflictsSeen.has(pairKey)) {
      _stemConflictsSeen.add(pairKey);
      const key = `${pair[0]}-${pair[1]}`;
      _stemConflicts[key] = {
        controller: stemId,
        controlled,
      };
    }
  }
}
export const STEM_CONFLICTS: Readonly<Record<string, StemConflictEntry>> = _stemConflicts;

// ---------------------------------------------------------------------------
// WUXING ENERGY FLOW (五行流動)
// ---------------------------------------------------------------------------

export const WUXING_ENERGY_FLOW = {
  generation: ELEMENT_CYCLES.generating,
  control: ELEMENT_CYCLES.controlling,
  scoring: {
    generating: {
      generator_loss: 7,
      receiver_gain: 10,
      count_increment: 0.25,
    },
    controlling: {
      controller_loss: 5,
      controlled_loss: 10,
      count_decrement: 0.25,
    },
  },
} as const;
